File: mobile/lib/core/order_pricing.dart
import '../models/service.dart' show ServiceItem;

const double checkoutGstRate = 0.18;

double _round2(double n) => (n * 100).roundToDouble() / 100;

/// Display INR with 2 decimal places (values are not truncated to integers).
String formatInrAmount(num value) {
  final amount = value is double ? value : value.toDouble();
  return amount.toStringAsFixed(2);
}

bool isGeneralServiceItem(ServiceItem service) {
  final cat = service.category ?? '';
  final name = service.name.toLowerCase();
  return cat == 'Periodic' ||
      cat == 'Services' ||
      name.contains('general service');
}

bool isGeneralServiceList(Iterable<ServiceItem> services) {
  return services.any(isGeneralServiceItem);
}

class OrderTotals {
  final double subtotal;
  final double discountAmount;
  final double discountedSubtotal;
  final double tax;
  final double total;

  const OrderTotals({
    required this.subtotal,
    required this.discountAmount,
    required this.discountedSubtotal,
    required this.tax,
    required this.total,
  });
}

OrderTotals calculateOrderTotals(
  double subtotal, [
  double discountAmount = 0,
  bool applyTax = true,
]) {
  final sub = _round2(subtotal < 0 ? 0 : subtotal);
  final disc = _round2(discountAmount < 0 ? 0 : discountAmount);
  final discountedSubtotal = _round2((sub - disc).clamp(0, double.infinity));
  final tax = applyTax ? _round2(discountedSubtotal * checkoutGstRate) : 0.0;
  final total = _round2(discountedSubtotal + tax);
  return OrderTotals(
    subtotal: sub,
    discountAmount: disc,
    discountedSubtotal: discountedSubtotal,
    tax: tax,
    total: total,
  );
}

File: backend/utils/orderPricing.js
/** Customer checkout GST rate (18% on subtotal after coupon discount). */
export const CHECKOUT_GST_RATE = 0.18;

const round2 = (n) => Math.round((Number(n) || 0) * 100) / 100;

export function isGeneralServiceItem(service) {
  if (!service) return false;
  const cat = service.category || '';
  const name = String(service.name || '').toLowerCase();
  return (
    cat === 'Periodic' ||
    cat === 'Services' ||
    name.includes('general service')
  );
}

export function isGeneralServiceList(services) {
  if (!Array.isArray(services) || services.length === 0) return false;
  return services.some(isGeneralServiceItem);
}

export async function isGeneralServiceByIds(serviceIds) {
  if (!Array.isArray(serviceIds) || serviceIds.length === 0) return false;
  try {
    const Service = (await import('../models/Service.js')).default;
    const services = await Service.find({ _id: { $in: serviceIds } }).lean();
    return isGeneralServiceList(services);
  } catch {
    return false;
  }
}

/** Prepaid checkout tax applies only to non–general-service bookings. */
export async function shouldApplyCheckoutGst(tempOrBookingData) {
  if (!tempOrBookingData) return false;
  if (tempOrBookingData.applyCheckoutGst === false) return false;
  if (isGeneralServiceList(tempOrBookingData.services)) return false;
  if (tempOrBookingData.serviceIds?.length) {
    return !(await isGeneralServiceByIds(tempOrBookingData.serviceIds));
  }
  return !!tempOrBookingData.requiresPaymentService;
}

/**
 * @param {number} subtotal - Services + pickup (before coupon and tax)
 * @param {number} [discountAmount=0] - Coupon discount applied to subtotal
 * @param {boolean} [applyTax=true] - When false (general service), tax is 0
 */
export function calculateOrderTotals(subtotal, discountAmount = 0, applyTax = true) {
  const sub = round2(Math.max(0, subtotal));
  const disc = round2(Math.max(0, discountAmount));
  const discountedSubtotal = round2(Math.max(0, sub - disc));
  const tax = applyTax ? round2(discountedSubtotal * CHECKOUT_GST_RATE) : 0;
  const total = round2(discountedSubtotal + tax);
  return {
    subtotal: sub,
    discountAmount: disc,
    discountedSubtotal,
    tax,
    total,
  };
}

export function mapCategoryToCouponServiceType(category) {
  if (!category) return null;
  const cat = String(category).toLowerCase();
  if (cat.includes('wash') || cat.includes('detailing')) return 'Car Wash';
  if (cat.includes('essential') || cat.includes('accessories')) return 'Essentials';
  if (cat.includes('tyre') || cat.includes('tire') || cat.includes('battery')) {
    return 'Tyres and Battery';
  }
  if (
    cat.includes('periodic') ||
    cat.includes('general') ||
    cat === 'services' ||
    cat === 'repair' ||
    cat === 'ac' ||
    cat === 'painting' ||
    cat === 'denting'
  ) {
    return 'General Service';
  }
  return null;
}

